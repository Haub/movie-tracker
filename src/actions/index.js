export const displayHanksMovies = (movies) => ({
  type: 'DISPLAY_HANKS_MOVIES',
  movies
})

export const updateFavorites = (movies) => ({
  type: 'UPDATE_FAVORITES',
  movies
export const addFavorite = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});

export const removeFavorite = (movieId) => ({
  type: 'REMOVE_FAVORITE',
  movieId
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
