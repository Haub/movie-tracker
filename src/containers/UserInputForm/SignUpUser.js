import React, {Component} from 'react';
import {connect} from 'react-redux';
import { NavLink } from 'react-router-dom';

import './Navigation.css';
import { userSignUp } from '../../Utilities/fetchApi';
import { loginUser, alertUser } from '../../actions';
import Alert from '../../containers/Alert';
 
export class SignUpUser extends Component {
  constructor() {
    super()
    this.state = {
      email: '',
      password: '',
      name: ''
    }
  }

  handleChange = (e) => {
    const {value, name} = e.target;
    this.setState({
      [name] : value
    });
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { login, history, alertUser } = this.props;
    const userInfo = await userSignUp(this.state);
    if (userInfo.alert) {
      return alertUser(userInfo.alert)
    }
    const newUser = {...this.state, id: userInfo.id}
    login(newUser);
    history.push('/user')
  }

  render() {
    const { email, password, name } = this.state;
    const { alertUser } = this.props;
    return (
      <div className="signup-cont">
        <h4>Sign Up to save your favorite Hanks movies!</h4>
        <form onSubmit={this.handleSubmit}>
          <input
            className='name-input'
            required
            placeholder='User Name'
            value={name}
            name='name'
            onChange={this.handleChange}
          />
          <input
            className='email-input'
            required
            type='email'
            placeholder='Email'
            value={email}
            name='email'
            onChange={this.handleChange}
          />
          <input
            className='password-input'
            required
            placeholder='Password'
            value={password}
            name='password'
            onChange={this.handleChange}
          />
          <button>Sign Up</button>
        </form>
        <Alert />
        <NavLink className="NavLink" to='/login' onClick={() => alertUser('')}>
          If you already have an account, Log In here!
        </NavLink>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  login: (user) => dispatch(loginUser(user)),
  alertUser: (message) => dispatch(alertUser(message))
})

export default connect(null, mapDispatchToProps)(SignUpUser);