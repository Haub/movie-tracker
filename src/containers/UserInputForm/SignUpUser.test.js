import React from 'react';
import { shallow } from 'enzyme';

import { SignUpUser, mapDispatchToProps } from './SignUpUser';
import { loginUser, alertUser } from '../../actions';


describe('SignUpUser', () => {
  describe('SignUpUser Component', () => {
    let wrapper;
    let login; 
    let history; 
    let alertUser;
    let e;

    beforeEach(() => {
      login = jest.fn(); 
      history = { push: jest.fn().mockImplementation(() => ({location:'/user'}))}
      alertUser = jest.fn();
      e = {preventDefault:jest.fn()}

      window.fetch = jest.fn().mockImplementation(() => (
        Promise.resolve({ json:() => Promise.resolve({id:1})})
      ))

      wrapper = shallow(
        <SignUpUser 
          login={login}
          history={history}
          alertUser={alertUser}
        />)
    })

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

    it('should call a reset on alert user on link to log in', () => {
      wrapper.find('NavLink').simulate('click');
      expect(alertUser).toHaveBeenCalled();
    });

    describe('handleChange', () => {
      it('should set state when invoked', () => {
        const wrapper = shallow(<SignUpUser />);
        let e = {target:{value:'foo', name:'email'}}

        wrapper.instance().handleChange(e);
        expect(wrapper.state().email).toEqual('foo');

        e = {target:{value:'foo', name:'password'}}
        wrapper.instance().handleChange(e);
        expect(wrapper.state().password).toEqual('foo');

        e = {target:{value:'foo', name:'username'}}
        wrapper.instance().handleChange(e);
        expect(wrapper.state().username).toEqual('foo');
      });
    });

    describe('handleSubmit', () => {
      beforeEach(() => {
        window.fetch = jest.fn().mockImplementation(() => (
          Promise.resolve({ json:() => Promise.resolve({id:1})})
        ))
        login = jest.fn(); 
        history = { push: jest.fn().mockImplementation(() => ({location:'/user'}))}
        alertUser = jest.fn();
        e = {preventDefault:jest.fn()}

        wrapper = shallow(
          <SignUpUser 
            login={login}
            history={history}
            alertUser={alertUser}
          />)
      })

      it('should call alertUser if fetch rejected', async () => {
        window.fetch = jest.fn().mockImplementation(() => (
          Promise.resolve({ json:() => Promise.reject()})
        ))
        await wrapper.instance().handleSubmit(e);
        expect(alertUser).toHaveBeenCalled();
      });

      it('should log in user if fetch passes', async () =>{
        await wrapper.instance().handleSubmit(e);
        expect(login).toHaveBeenCalled();
      });

      it('should invoke alertUser if userInfo has alert', () => {
        const mockAlertUser = jest.fn();
        const mockEvent = { preventDefault: jest.fn() }
        wrapper = shallow(<SignUpUser alertUser={mockAlertUser} />);
        wrapper.setState({userInfo: {alert: 'bingo'}});
        wrapper.instance().handleSubmit(mockEvent)

        expect(mockAlertUser).toHaveBeenCalled()
      });
    });
  });
  describe('mapDispatchToProps', () => {
    it.skip('should dispatch loginUser action when login is invoked', () => {
      const mockUser = {
        name: 'wil',
        email: 'a@a',
        password: 'wil'
      };
      const mockDispatch = jest.fn();
      const actionToDispatch = loginUser(mockUser)
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.login(mockUser)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it.skip('should dispatch alertUser action when alertUser is invoked', () => {
      const mockMessage = { alert: 'Email has already been taken.' };
      const mockDispatch = jest.fn();
      const actionToDispatch = alertUser(mockMessage);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.alertUser(mockMessage);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});