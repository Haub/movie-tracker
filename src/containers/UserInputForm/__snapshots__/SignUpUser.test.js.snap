// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpUser SignUpUser Component should match snapshot 1`] = `
<div
  className="signup-cont"
>
  <h4>
    Sign Up to save your favorite Hanks movies!
  </h4>
  <form
    onSubmit={[Function]}
  >
    <input
      className="name-input"
      name="name"
      onChange={[Function]}
      placeholder="User Name"
      required={true}
      value=""
    />
    <input
      className="email-input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <input
      className="password-input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      value=""
    />
    <button>
      Sign Up
    </button>
  </form>
  <Connect(Component) />
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="NavLink"
    onClick={[Function]}
    to="/login"
  >
    If you already have an account, Log In here!
  </NavLink>
</div>
`;
